With this patch it starts to compile on darwin 🎉 :

```patch
diff --git a/pkgs/misc/emulators/retroarch/cores.nix b/pkgs/misc/emulators/retroarch/cores.nix
index 4a8ca8713e3..6e970098a4b 100644
--- a/pkgs/misc/emulators/retroarch/cores.nix
+++ b/pkgs/misc/emulators/retroarch/cores.nix
@@ -549,7 +549,9 @@ in with stdenv.lib.licenses;
     description = "Port of MAME to libretro";
     license = gpl2Plus;
 
-    extraBuildInputs = [ alsaLib libGLU libGL portaudio python27 xorg.libX11 ];
+    extraBuildInputs = [ libGLU libGL portaudio python27 xorg.libX11 ]
+      ++ stdenv.lib.optional stdenv.isLinux alsaLib;
+
     postPatch = ''
       # Prevent the failure during the parallel building of:
       # make -C 3rdparty/genie/build/gmake.linux -f genie.make obj/Release/src/host/lua-5.3.0/src/lgc.o
@@ -618,7 +620,7 @@ in with stdenv.lib.licenses;
     description = "Port of MAME ~2015 to libretro";
     license = gpl2Plus;
     extraNativeBuildInputs = [ python27 ];
-    extraBuildInputs = [ alsaLib ];
+    extraBuildInputs = stdenv.lib.optional stdenv.isLinux alsaLib;
     makefile = "Makefile";
   };
 
@@ -639,7 +641,7 @@ in with stdenv.lib.licenses;
     description = "Port of MAME ~2016 to libretro";
     license = gpl2Plus;
     extraNativeBuildInputs = [ python27 ];
-    extraBuildInputs = [ alsaLib ];
+    extraBuildInputs = stdenv.lib.optional stdenv.isLinux alsaLib;
     postPatch = ''
       # Prevent the failure during the parallel building of:
       # make -C 3rdparty/genie/build/gmake.linux -f genie.make obj/Release/src/host/lua-5.3.0/src/lgc.o
diff --git a/pkgs/misc/emulators/retroarch/default.nix b/pkgs/misc/emulators/retroarch/default.nix
index bfc2c338769..9b2a1db9462 100644
--- a/pkgs/misc/emulators/retroarch/default.nix
+++ b/pkgs/misc/emulators/retroarch/default.nix
@@ -32,7 +32,8 @@ stdenv.mkDerivation rec {
     rev = "v${version}";
   };
 
-  nativeBuildInputs = [ pkgconfig wayland ]
+  nativeBuildInputs = [ pkgconfig ]
+                      ++ optional stdenv.isLinux wayland
                       ++ optional withVulkan makeWrapper;
 
   buildInputs = [ ffmpeg_3 freetype libxml2 libGLU libGL python3 SDL2 which ]
```

I moved all alsaLibs to Linux only but did not test if this is required.

But then it fails to compile because it can't find gcc

```
no configure script, doing nothing
building
build flags: -j4 -l4 SHELL=/nix/store/smimdn17mp2ih6nfxwl78mswwdkmcm56-bash-4.4-p23/bin/bash platform=osx ARCH=x86_64
make REGENIE=1 VERBOSE=1 NOWERROR=1 OSD="retro" PYTHON_EXECUTABLE=python CONFIG=libretro LIBRETRO_OS="osx" ARCH="" LIBRETRO_CPU="x86_64"
make[1]: Entering directory '/private/var/folders/r4/k85s52jx3555ckj0wty615s00000gp/T/nix-build-libretro-mame2016-2020-03-06.drv-0/mame2016-libretro-02987af'
GCC  detected
make -R verbose=1 -C 3rdparty/genie/build/gmake.darwin -f genie.make
mkdir -p "build/generated/mame/layout/"
mkdir -p "build/generated/mame/mame/"
mkdir -p "build/generated/mame/drivers/"
make[2]: Entering directory '/private/var/folders/r4/k85s52jx3555ckj0wty615s00000gp/T/nix-build-libretro-mame2016-2020-03-06.drv-0/mame2016-libretro-02987af/3rdparty/genie/build/gmake.darwin'
Creating obj/Release
Converting translation language/Afrikaans/strings.po...
mkdir -p "obj/Release"
python scripts/build/msgfmt.py --output-file language/Afrikaans/strings.mo language/Afrikaans/strings.po
Converting translation language/Albanian/strings.po...
Creating obj/Release/src/host/lua-5.3.0/src
python scripts/build/msgfmt.py --output-file language/Albanian/strings.mo language/Albanian/strings.po
mkdir -p "obj/Release/src/host/lua-5.3.0/src"
Creating obj/Release/src/host
mkdir -p "obj/Release/src/host"
os_getcwd.c
gcc   -MMD -MP -MP -DNDEBUG -DLUA_COMPAT_MODULE -DLUA_USE_MACOSX -I../../src/host/lua-5.3.0/src  -Wall -Wextra -Os -mmacosx-version-min=10.4  -o "obj/Release/src/host/os_getcwd.o" -c "../../src/host/os_getcwd.c"
/nix/store/smimdn17mp2ih6nfxwl78mswwdkmcm56-bash-4.4-p23/bin/bash: gcc: command not found
make[2]: *** [genie.make:406: obj/Release/src/host/os_getcwd.o] Error 127
make[2]: *** Waiting for unfinished jobs....
make[2]: Leaving directory '/private/var/folders/r4/k85s52jx3555ckj0wty615s00000gp/T/nix-build-libretro-mame2016-2020-03-06.drv-0/mame2016-libretro-02987af/3rdparty/genie/build/gmake.darwin'
make[1]: *** [makefile:1453: 3rdparty/genie/bin/darwin/genie] Error 2
make[1]: *** Waiting for unfinished jobs....
Converting translation language/Arabic/strings.po...
python scripts/build/msgfmt.py --output-file language/Arabic/strings.mo language/Arabic/strings.po
make[1]: Leaving directory '/private/var/folders/r4/k85s52jx3555ckj0wty615s00000gp/T/nix-build-libretro-mame2016-2020-03-06.drv-0/mame2016-libretro-02987af'
make: *** [Makefile.libretro:197: build] Error 2
builder for '/nix/store/w1nz7ldp5vh1x1q9fddc6v6ha2c0bf8z-libretro-mame2016-2020-03-06.drv' failed with exit code 2
error: build of '/nix/store/w1nz7ldp5vh1x1q9fddc6v6ha2c0bf8z-libretro-mame2016-2020-03-06.drv' failed
```

--- a/pkgs/misc/emulators/retroarch/cores.nix
+++ b/pkgs/misc/emulators/retroarch/cores.nix
@@ -630,7 +630,6 @@ in with stdenv.lib.licenses;
       rev = "02987af9b81a9c3294af8fb9d5a34f9826a2cf4d";
       sha256 = "0gl7irmn5d8lk7kf484vgw6kb325fq4ghwsni3il4nm5n2a8yglh";
     };
-    broken = true;
     patches = [
       (fetchpatch {
         name = "fix_mame_build_on_make-4.3.patch";
@@ -640,6 +639,7 @@ in with stdenv.lib.licenses;
     ];
     description = "Port of MAME ~2016 to libretro";
     license = gpl2Plus;
+    broken = stdenv.isDarwin; # https://github.com/NixOS/nixpkgs/pull/102078#issuecomment-719535596
     extraNativeBuildInputs = [ python27 ];
     extraBuildInputs = [ alsaLib ];
     postPatch = ''

--- a/pkgs/misc/emulators/retroarch/cores.nix
+++ b/pkgs/misc/emulators/retroarch/cores.nix
@@ -1,4 +1,4 @@
-{ stdenv, fetchgit, fetchFromGitHub, fetchFromGitLab, cmake, pkgconfig, makeWrapper, python27, python37, retroarch
+{ stdenv, fetchgit, fetchFromGitHub, fetchFromGitLab, fetchpatch, cmake, pkgconfig, makeWrapper, python27, python37, retroarch
 , alsaLib, fluidsynth, curl, hidapi, libGLU, gettext, glib, gtk2, portaudio, SDL, SDL_net, SDL2, SDL2_image, libGL
 , ffmpeg_3, pcre, libevdev, libpng, libjpeg, libzip, udev, libvorbis, snappy, which, hexdump
 , miniupnpc, sfml, xorg, zlib, nasm, libpcap, boost, icu, openssl
@@ -631,6 +631,13 @@ in with stdenv.lib.licenses;
       sha256 = "0gl7irmn5d8lk7kf484vgw6kb325fq4ghwsni3il4nm5n2a8yglh";
     };
     broken = true;
+    patches = [
+      (fetchpatch {
+        name = "fix_mame_build_on_make-4.3.patch";
+        url = "https://github.com/libretro/mame2016-libretro/commit/5874fae3d124f5e7c8a91634f5473a8eac902e47.patch";
+        sha256 = "061f1lcm72glksf475ikl8w10pnbgqa7049ylw06nikis2qdjlfn";
+      })
+    ];
     description = "Port of MAME ~2016 to libretro";
     license = gpl2Plus;
     extraNativeBuildInputs = [ python27 ];

